apps/src/calc/equationSet.js

```diff
@@ -0,0 +1,312 @@
+ */
+EquationSet.prototype.addEquation_ = function (equation) {
+  if (!equation.name) {
+    if (this.compute_) {
+      throw new Error('compute expression already exists');
+    }
+    this.compute_ = equation;
+  } else {
+    if (this.getEquation(equation.name)) {
+      throw new Error('equation already exists: ' + equation.name);
+    }
+    this.equations_.push(equation);
+  }
+};
+
+/**
+ * Get an equation by name, or compute equation if name is null
+ * @returns {Equation} Equation of that name if it exists, null otherwise.
+ */
+EquationSet.prototype.getEquation = function (name) {
+  if (name === null) {
+    return this.computeEquation();
+  }
+  for (var i = 0; i < this.equations_.length; i++) {
+    if (this.equations_[i].name === name) {
+      return this.equations_[i];
+    }
+  }
+  return null;
+};
+
+/**
+ * @returns the compute equation if there is one
+ */
+EquationSet.prototype.computeEquation = function () {
+  return this.compute_;
+};
+
+/**
+ * @returns true if EquationSet has at least one variable or function.
+ */
+EquationSet.prototype.hasVariablesOrFunctions = function () {
+  return this.equations_.length > 0;
+};
+
+/**
+ * If the EquationSet has exactly one function and no variables, returns that
+ * equation. If we have multiple functions or one function and some variables,
+ * it returns null
+ */
+EquationSet.prototype.singleFunction_ = function () {
+  if (this.equations_.length === 1 && this.equations_[0].isFunction()) {
+    return this.equations_[0];
+  }
+
+  return null;
+};
+
+/**
+ * External callers only care about whether or not we have a single function
+ */
+ EquationSet.prototype.hasSingleFunction = function () {
+   return this.singleFunction_() !== null;
+ };
+
+/**
+ * Are two EquationSets identical? This is considered to be true if their
+ * compute expressions are identical and all of their equations have the same
+ * names and identical expressions.
+ */
+EquationSet.prototype.isIdenticalTo = function (otherSet) {
+  if (this.equations_.length !== otherSet.equations_.length) {
+    return false;
+  }
+
+  var otherCompute = otherSet.computeEquation().expression;
+  if (!this.compute_.expression.isIdenticalTo(otherCompute)) {
+    return false;
+  }
+
+  for (var i = 0; i < this.equations_.length; i++) {
+    var thisEquation = this.equations_[i];
+    var otherEquation = otherSet.getEquation(thisEquation.name);
+    if (!otherEquation ||
+        !thisEquation.expression.isIdenticalTo(otherEquation.expression)) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+/**
+ * Returns a list of equations (vars/functions) sorted by name.
+ */
+EquationSet.prototype.sortedEquations = function () {
+  // TODO - this has side effects, do i care?
+  // sort by name. note - this sorts in place
+  this.equations_.sort(function (a, b) {
+    return a.name.localeCompare(b.name);
+  });
+
+  // append compute expression with name null
+  // return this.equations_.concat(this.compute_);
```

Comment:
Yeah, something I ended up not needing - callers are responsible for dealing with the compute equation if that want that in addition. I should probably also make that more explicit the function.